src/features/events/store.ts

```diff
@@ -166,6 +166,26 @@ const eventsSlice = createSlice({
       item.data = event;
       item.isLoading = false;
       item.loaded = new Date().toISOString();
+
+      if (event.campaign) {
+        if (!state.eventsByCampaignId[event.campaign.id]) {
+          state.eventsByCampaignId[event.campaign.id] =
+            remoteList<ZetkinEvent>();
+          state.eventsByCampaignId[event.campaign.id].items.push(
+            remoteItem(event.id, { data: event })
+          );
+        }
+
+        const eventItem = state.eventsByCampaignId[
+          event.campaign.id
+        ].items.find((item) => item.id == event.id);
+        if (eventItem) {
+          eventItem.data = { ...eventItem.data, ...event };
+          eventItem.mutating = [];
+        }
+        state.eventsByCampaignId[event.campaign!.id].loaded =
```

Comment:
You're using `event.campaign!` with the non-null assertion operator (`!`). I don't like that, because it means you're telling typescript you know better than typescript, which may or may not be true.

It's probably true that you know better than typescript in this place, right now (because you did an if statement on line 170 that should mean that `event.campaign` is not null). But what happens if some other coder in the future comes and changes that if statement on line 170, so that `event.campaign` no longer guaranteed to be non-null here on line 186? The code would start crashing.

It's better to not rely on "tricking" typescript. That will make the code more reliable long-term.

In this case, the reason why typescript is telling you that `event.campaign` might be `null` is because it's being careful. The `event.campaign` expression is not constant. It could change. In this case, Typescript is being _too_ careful, because I can tell from looking at the code that it could never have changed between line 170 and here. 

But Typescript is being careful because javascript control flow can be difficult to follow. Functions can change variables outside the functions, and there can be multiple levels of closures that are difficult for Typescript to follow. So typescript will act like this unles you store the value in a constant variable that typescript knows won't change.

In this case, change line 169 to do `const campaign = event.campaign`, and use `campaign` both in the if statement on line 170 and here on line 186 and everywhere else in this function.

Reply:
Really weird behavior of typescript.. Thank you so much for the explanation!